File: rb/12.rb
def prime_factors n
  factors = []

  if n % 2 == 0
    factors << 2
    factors << 2 while (n /= 2) % 2 == 0
  end

  p = 3
  while n > 1
    if n % p == 0
      factors << p
      factors << p while (n /= p) % p == 0
    end
    p += 2
  end

  factors
end

def divisors n
  pf = prime_factors n
  pf.uniq.map { |f| pf.count(f) + 1 }.reduce(1, :*)
end

n, p = 1, 1
n += (p += 1) until divisors(n) > 500
p n

File: py/12.py
def prime_factors(n):
    factors = []

    if n % 2 == 0:
        while n % 2 == 0:
            factors.append(2)
            n /= 2

    p = 3
    while n > 1:
        if n % p == 0:
            while n % p == 0:
                factors.append(p)
                n /= p
        p += 2

    return factors

def divisors(n):
    pf = prime_factors(n)
    return reduce(lambda a, b: a * b, map(lambda f: pf.count(f) + 1, set(pf)), 1)

n, p = 1, 1
while divisors(n) < 500:
    p += 1
    n += p

print n
